File: scripts/generate_articles.py
#!/usr/bin/env python3
from __future__ import annotations

import csv
import json
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
INPUT_CSV = ROOT / 'data' / 'items.csv'
OUTPUT_TS = ROOT / 'src' / 'data' / 'articles.generated.ts'

CATEGORY_MAP = {
    'betrieb': 'betrieb',
    'infrastruktur': 'infrastruktur',
    'plattform': 'plattform',
    'zugang': 'zugang',
}


def normalize_category(raw: str) -> str:
    compact = ''.join(raw.split()).lower()
    for key, value in CATEGORY_MAP.items():
        if key in compact:
            return value
    return 'plattform'


def parse_tags(raw: str) -> list[str]:
    if not raw:
        return []
    return [tag.strip() for tag in raw.split(',') if tag.strip()]


def normalize_logo(raw: str) -> str:
    cleaned = (raw or '').strip()
    if not cleaned:
        return ''
    if cleaned.startswith('http://') or cleaned.startswith('https://'):
        return cleaned
    return '/' + cleaned.lstrip('/')


def main() -> None:
    rows: list[dict[str, object]] = []
    with INPUT_CSV.open(newline='', encoding='utf-8') as fh:
        reader = csv.DictReader(fh)
        for index, row in enumerate(reader, start=1):
            rows.append(
                {
                    'id': str(index),
                    'name': (row.get('name') or '').strip(),
                    'category': normalize_category(row.get('category') or ''),
                    'description': (row.get('description') or '').strip(),
                    'logo': normalize_logo(row.get('logo') or ''),
                    'tags': parse_tags(row.get('tag') or ''),
                    'featured': False,
                }
            )

    generated_at = datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace('+00:00', 'Z')

    lines: list[str] = [
        '// GENERATED FILE - DO NOT EDIT MANUALLY',
        f'// Generated from: {INPUT_CSV.relative_to(ROOT).as_posix()}',
        f'// Generated at: {generated_at}',
        '',
        "import { Article } from '../types';",
        '',
        "const ASSET_BASE_URL = (import.meta.env.VITE_ASSET_BASE_URL ?? '').replace(/\\/+$/, '');",
        '',
        'function withAssetBaseUrl(logo?: string): string | undefined {',
        '	if (!logo || /^https?:\\/\\//.test(logo)) return logo;',
        "	const normalizedPath = `/${logo.replace(/^\\/+/, '')}`;",
        '	return ASSET_BASE_URL ? `${ASSET_BASE_URL}${normalizedPath}` : normalizedPath;',
        '}',
        '',
        'const RAW_ARTICLES: Article[] = [',
    ]

    for row in rows:
        lines.extend(
            [
                '\t{',
                f"\t\tid: {json.dumps(row['id'], ensure_ascii=False)},",
                f"\t\tname: {json.dumps(row['name'], ensure_ascii=False)},",
                f"\t\tcategory: {json.dumps(row['category'], ensure_ascii=False)},",
                f"\t\tdescription: {json.dumps(row['description'], ensure_ascii=False)},",
                f"\t\tlogo: {json.dumps(row['logo'], ensure_ascii=False)},",
                f"\t\ttags: {json.dumps(row['tags'], ensure_ascii=False)},",
                '\t\tfeatured: false,',
                '\t},',
            ]
        )

    lines.extend(
        [
            '];',
            '',
            'export const ARTICLES: Article[] = RAW_ARTICLES.map((article) => ({',
            '\t...article,',
            '\tlogo: withAssetBaseUrl(article.logo),',
            '}));',
            '',
        ]
    )

    OUTPUT_TS.write_text('\n'.join(lines), encoding='utf-8')
    print(f'Generated {OUTPUT_TS.relative_to(ROOT)} from {INPUT_CSV.relative_to(ROOT)}')


if __name__ == '__main__':
    main()

File: scripts/generate-articles.mjs
#!/usr/bin/env node

import { readFileSync, writeFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..');
const INPUT_CSV = join(ROOT, 'data', 'items.csv');
const LOGO_URLS_JSON = join(ROOT, 'src', 'data', 'logo-urls.json');
const OUTPUT_TS = join(ROOT, 'src', 'data', 'articles.generated.ts');

const CATEGORIES = new Set(['betrieb', 'infrastruktur', 'plattform', 'zugang']);
const FALLBACK_LOGO = '/assets/broken-logo.svg';

function parseCSV(csv) {
	const [headerLine, ...dataLines] = csv.trim().split('\n');
	const headers = parseLine(headerLine);

	return dataLines.filter(Boolean).map((line) => {
		const values = parseLine(line);
		return Object.fromEntries(headers.map((h, i) => [h, values[i] || '']));
	});
}

function parseLine(line) {
	const fields = [];
	let current = '';
	let quoted = false;

	for (let i = 0; i < line.length; i++) {
		const ch = line[i];
		if (ch === '"') {
			if (quoted && line[i + 1] === '"') {
				current += '"';
				i++;
			} else {
				quoted = !quoted;
			}
		} else if (ch === ',' && !quoted) {
			fields.push(current.trim());
			current = '';
		} else {
			current += ch;
		}
	}
	fields.push(current.trim());
	return fields;
}

function toArticle(row, index, logoUrls) {
	const category = [...CATEGORIES].find((c) => (row.category || '').replace(/\s/g, '').toLowerCase().includes(c)) || 'plattform';
	const csvLogo = (row.logo || '').trim();
	const name = (row.name || '').trim();
	const tags = (row.tag || '')
		.split(',')
		.map((t) => t.trim())
		.filter(Boolean);

	// 1. Use CSV logo if it's an external URL
	let logo = '';
	if (csvLogo && /^https?:\/\//.test(csvLogo)) {
		logo = csvLogo;
	} else if (csvLogo && !csvLogo.startsWith('http')) {
		// 2. Use CSV logo if it's a local path
		logo = '/' + csvLogo.replace(/^\/+/, '');
	}

	// 3. Try external logo URL from logo-urls.json
	if (!logo && logoUrls[name]) {
		const logoEntry = logoUrls[name];
		if (logoEntry.verified && logoEntry.url) {
			logo = logoEntry.url;
		}
	}

	// 4. Fallback to broken-logo.svg if nothing found
	if (!logo) {
		logo = FALLBACK_LOGO;
	}

	return {
		id: String(index + 1),
		name,
		category,
		description: (row.description || '').trim(),
		logo,
		tags,
		featured: false,
	};
}

// Load logo-urls.json for fallback/override
let logoUrls = {};
try {
	logoUrls = JSON.parse(readFileSync(LOGO_URLS_JSON, 'utf-8'));
} catch {
	console.warn('⚠️  Could not load logo-urls.json, will use CSV logos and fallback only');
}

const articles = parseCSV(readFileSync(INPUT_CSV, 'utf-8')).map((row, idx) => toArticle(row, idx, logoUrls));

const articlesJson = articles.map((a) => `\t${JSON.stringify(a)},`).join('\n');

const output = `// GENERATED FILE - DO NOT EDIT MANUALLY
// Generated from: data/items.csv + src/data/logo-urls.json
// Generated at: ${new Date().toISOString()}

import { Article } from '../types';

const ASSET_BASE_URL = (import.meta.env.VITE_ASSET_BASE_URL ?? '').replace(/\\/+$/, '');

function withAssetBaseUrl(logo?: string): string | undefined {
\tif (!logo || /^https?:\\/\\//.test(logo)) return logo;
\tconst normalizedPath = \`/\${logo.replace(/^\\/+/, '')}\`;
\treturn ASSET_BASE_URL ? \`\${ASSET_BASE_URL}\${normalizedPath}\` : normalizedPath;
}

const RAW_ARTICLES: Article[] = [
${articlesJson}
];

export const ARTICLES: Article[] = RAW_ARTICLES.map((article) => ({
\t...article,
\tlogo: withAssetBaseUrl(article.logo),
}));
`;

writeFileSync(OUTPUT_TS, output, 'utf-8');

// Count logo types
const logoCounts = articles.reduce(
	(acc, a) => {
		if (a.logo === FALLBACK_LOGO) {
			acc.fallback++;
		} else if (/^https?:\/\//.test(a.logo)) {
			acc.external++;
		} else {
			acc.local++;
		}
		return acc;
	},
	{ external: 0, local: 0, fallback: 0 },
);

console.log(`Generated src/data/articles.generated.ts (${articles.length} articles)`);
console.log(`  📍 External URLs: ${logoCounts.external}`);
console.log(`  📁 Local paths: ${logoCounts.local}`);
console.log(`  ⚠️  Fallback logos: ${logoCounts.fallback}`);
